test/bin/scenario.sh

```diff
@@ -1069,6 +1068,187 @@ EOF
+        fi
+        cd "${tests_repo_dir}"
+    fi
+    record_junit "${vmname}" "clone_tests_repo" "OK"
+
+    # Build the test binary
+    echo "Building openshift-tests-private binary..."
+    local test_binary="./bin/extended-platform-tests"
+
+    if [[ -f "Makefile" ]]; then
+        if ! make -j"$(nproc)" build; then
+            record_junit "${vmname}" "build_test_binary" "FAILED"
+            return 1
+        fi
+    elif [[ -f "go.mod" ]]; then
+        if ! go build -o "${test_binary}" ./cmd/extended-platform-tests; then
+            record_junit "${vmname}" "build_test_binary" "FAILED"
+            return 1
+        fi
+    else
+        error "Unable to determine build method for tests repository"
+        record_junit "${vmname}" "build_test_binary" "FAILED"
+        return 1
+    fi
+
+    if [[ ! -f "${test_binary}" ]]; then
+        error "Test binary not found after build: ${test_binary}"
+        record_junit "${vmname}" "build_test_binary" "FAILED"
+        return 1
+    fi
+
+    record_junit "${vmname}" "build_test_binary" "OK"
+
+    # Return the test binary path
+    echo "${tests_repo_dir}/${test_binary}"
+}
+
+# Implementation of Gingko tests
+run_gingko_tests() {
+    local vmname="${1}"
+    shift
+
+    local full_vmname
+    full_vmname="$(full_vm_name "${vmname}")"
+    if [[ -n "${RUN_HOST_OVERRIDE}" ]]; then
+        vmname="${RUN_HOST_OVERRIDE}"
+        full_vmname="$(full_vm_name "${vmname}")"
+        ip=$(get_vm_ip "${full_vmname}")
+        set_vm_property "${vmname}" "ip" "${ip}"
+        set_vm_property "${vmname}" "ssh_port" "22"
+        set_vm_property "${vmname}" "api_port" "6443"
+        set_vm_property "${vmname}" "lb_port" "5678"
+    fi
+
+    local original_dir
+    original_dir="$(pwd)"
+
+    # Check/install oc
+    if ! command -v oc &> /dev/null; then
+        "${ROOTDIR}/scripts/fetch_tools.sh" "oc" || {
+            record_junit "${vmname}" "oc_installed" "FAILED"
+            exit 1
+        }
+    fi
+
+    local test_binary
+    if ! test_binary="$(build_ginkgo_binary "${vmname}")"; then
+        cd "${original_dir}"
+        exit 1
+    fi
+
+    # Set up test environment variables
+    local -r test_results_dir="${SCENARIO_INFO_DIR}/${SCENARIO}/gingko-results"
+    mkdir -p "${test_results_dir}"
+
+    # Set up kubeconfig for tests
+    local -r kubeconfig_file="${test_results_dir}/kubeconfig"
+    local -r vm_ip=$(get_vm_property "${vmname}" "ip")
+    local -r full_vmname="$(full_vm_name "${vmname}")"
+    local -r vm_hostname="${full_vmname/./-}"
+
+    # Get kubeconfig
+    run_command_on_vm "${vmname}" "sudo cp /var/lib/microshift/resources/kubeadmin/${vm_hostname}/kubeconfig /home/redhat/kubeconfig-${vm_hostname}" || {
+        error "Failed to copy kubeconfig from /var/lib/microshift/resources/kubeadmin/${vm_hostname}/kubeconfig"
+        record_junit "${vmname}" "setup_kubeconfig" "FAILED"
+        cd "${original_dir}"
```

Comment:
Note that the dir restoration is not necessary before fatal exit, only on successful return from the function